File: backend/app/models/place.py
from sqlmodel import SQLModel, Field, Relationship, Column
from typing import Optional, List
from datetime import datetime
from sqlalchemy.dialects.postgresql import JSONB


class Place(SQLModel, table=True):
    """Destinations, restaurants, hotels, attractions"""
    __tablename__ = "places"
    
    id: str = Field(primary_key=True)  # External ID (Google Place ID, etc.)
    name: str = Field(index=True, max_length=200)
    description: Optional[str] = None
    category: Optional[str] = Field(default=None, index=True, max_length=50)
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    address: Optional[str] = None
    phone: Optional[str] = Field(default=None, max_length=50)
    website: Optional[str] = None
    rating: Optional[float] = Field(default=None, ge=0, le=5)
    price_level: Optional[int] = Field(default=None, ge=1, le=4)
    opening_hours: dict = Field(default_factory=dict, sa_column=Column(JSONB, default=dict))
    photos: List[str] = Field(default_factory=list, sa_column=Column(JSONB, default=list))
    place_metadata: dict = Field(default_factory=dict, sa_column=Column(JSONB, default=dict))
    source: Optional[str] = Field(default=None, max_length=50)  # 'brave_search', 'manual', etc.
    last_updated: datetime = Field(default_factory=datetime.utcnow)
    created_at: datetime = Field(default_factory=datetime.utcnow)


class SavedPlace(SQLModel, table=True):
    """Places saved by users"""
    __tablename__ = "saved_places"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: str = Field(foreign_key="user_profiles.id", index=True)
    place_id: str = Field(foreign_key="places.id", index=True)
    notes: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


# Response models
class PlacePublic(SQLModel):
    id: str
    name: str
    category: Optional[str]
    latitude: float
    longitude: float
    rating: Optional[float]


class PlaceDetail(PlacePublic):
    description: Optional[str]
    address: Optional[str]
    phone: Optional[str]
    website: Optional[str]
    price_level: Optional[int]
    opening_hours: dict
    photos: List[str]

File: frontend/atla/src/types/place.ts
/**
 * Place Type Definitions
 * 
 * These types match the backend SQLModel/Pydantic models exactly.
 * Used for type-safe API communication and component props.
 */

/**
 * Place (destination, restaurant, hotel, attraction)
 * Matches backend: app/models/place.py - Place
 */
export interface Place {
  id: string  // External ID (Google Place ID, etc.)
  name: string
  description?: string
  category?: string
  latitude: number
  longitude: number
  address?: string
  phone?: string
  website?: string
  rating?: number  // 0-5
  price_level?: number  // 1-4
  opening_hours: Record<string, unknown>  // JSONB
  photos: string[]  // Array of photo URLs
  place_metadata: Record<string, unknown>  // JSONB - renamed from 'metadata' to avoid conflict
  source?: string  // 'brave_search', 'manual', etc.
  last_updated: string  // ISO date string
  created_at: string
}

/**
 * Public place view (minimal info for listings)
 * Matches backend: PlacePublic Pydantic model
 */
export interface PlacePublic {
  id: string
  name: string
  category?: string
  latitude: number
  longitude: number
  rating?: number
}

/**
 * Place detail (full info)
 * Matches backend: PlaceDetail Pydantic model
 */
export interface PlaceDetail extends PlacePublic {
  description?: string
  address?: string
  phone?: string
  website?: string
  price_level?: number
  opening_hours: Record<string, unknown>
  photos: string[]
}

/**
 * Saved place (user's bookmarked places)
 * Matches backend: app/models/place.py - SavedPlace
 */
export interface SavedPlace {
  id: number
  user_id: string
  place_id: string
  notes?: string
  created_at: string
}

/**
 * Saved place with full place details
 * Used when fetching user's saved places list
 */
export interface SavedPlaceWithDetails extends SavedPlace {
  place: Place
}

/**
 * Search places request parameters
 */
export interface SearchPlacesParams {
  query: string
  lat?: number
  lng?: number
  radius?: number  // in kilometers
  category?: string
  limit?: number
}

/**
 * Category filter options for places
 */
export type PlaceCategory = 
  | 'attraction'
  | 'restaurant'
  | 'accommodation'
  | 'transport'
  | 'shopping'
  | 'nature'
  | 'cultural'
  | 'adventure'
  | 'relaxation'
  | string  // Allow custom categories

/**
 * Place search result item
 * Simplified version for search results
 */
export interface PlaceSearchResult {
  id: string
  name: string
  category?: string
  latitude: number
  longitude: number
  address?: string
  rating?: number
  price_level?: number
  thumbnail?: string  // First photo or placeholder
}

/**
 * Opening hours structure
 * Common format for place opening hours
 */
export interface OpeningHours {
  [day: string]: {
    open: string  // "09:00"
    close: string // "18:00"
    closed?: boolean
  }
}

/**
 * Place metadata structure
 * Flexible metadata storage
 */
export interface PlaceMetadata {
  google_place_id?: string
  tripadvisor_id?: string
  tags?: string[]
  amenities?: string[]
  accessibility?: string[]
  [key: string]: unknown
}
